server/src/com/cloud/storage/snapshot/SnapshotManagerImpl.java

```diff
@@ -530,6 +534,13 @@ public boolean deleteSnapshot(long snapshotId) {
             return false;
         }
 
+        List<VolumeDetailVO> volumeCheck;
+        volumeCheck = _volumeDetailsDaoImpl.findDetails("SNAPSHOT_ID", String.valueOf(snapshotId), null);
+
+        if (volumeCheck.size() > 0) {
+            throw new InvalidParameterValueException("Unable to perform delete operation, Snapshot with id: " + snapshotId + " is in use  ");
+        }
+
```

Comment:
When a volume created from snapshot, we are not maintaining volume in relation to snapshot details anywhere in database. It's been recently integrated and information stored in VOLUME_DETAILS table.

Data/Records in this table will be persist till volume creation operation finish(success/failure). Once volume operation done this record will be removed. I have not come across alternatives other than this approach.